proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,566 @@
+  considered an overpay case. The payer balance is checked for
+  `other fees + PRAW fees`.
+  1. The payer does not have enough balance: Transaction fails with an error
+    `Insufficient Balance` and the transaction is not even scheduled for
+    execution.
+  2. The payer has enough balance then the transaction is executed and PRAW
+    fees paid are transferred back to the payer in any case.
+  
+  Note in this case
+    even if there are no PRAW fees involved the payer balance is checked
+    against PRAW fees.
+
+#### PRAW fees are enabled
+
+* Fees not paid case:
+
+  A payer does not include `SetProgramRebatableAccountWriteFees` in the
+  transaction. The transaction includes one or more accounts with PRAW fees.
+  Then the transaction is failed with an error
+  `ProgramRebatableAccountWriteFeesNotPaid`. The program is not executed at all.
+  The payer ends up paying only other fees.
+
+* Fees paid no rebates case:
+
+  A payer includes instruction `SetProgramRebatableAccountWriteFees(100)` in the
+  transaction. There is an account `accA` which is write-locked by the
+  transaction and it has an PRAW fee of `100` lamports. Consider that the
+  program does not have any rebate mechanism. Then in any case (execution fails
+  or succeeds) `accA` will receive `100` lamports. The payer will end up paying
+  `other fees` + `100` lamports.
+
+* Fees paid full rebates case:
+
+  A payer includes instruction `SetProgramRebatableAccountWriteFees(100)` in the
+  transaction. There is an accounts `accA` which is write-locked by the
+  transaction and it has an PRAW fees fee of `100` lamports. Consider during
+  execution the program will rebate the PRAW fees fee on the account. Then payer
+  should have a minimum balance of `other fees` + `100` lamports to execute the
+  transaction. After successful execution of the transaction, the `100` lamports
+  will be rebated by the program and then Solana runtime will transfer them back
+  to the payer. So the payer will finally end up paying `other fees` only.
+
+* Fees paid multiple partial rebates case:
+
+  A payer includes instruction `SetProgramRebatableAccountWriteFees(200)` in the
+  transaction. The transaction has three instructions (`Ix1`, `Ix2`, `Ix3`).
+  There are accounts (`accA`, `accB`) that are write-locked by the transaction
+  and each of them has an PRAW fees fee of `100` lamports. Lets consider `Ix1`
+  rebates 25 lamports on both accounts, `Ix2` rebates 75 lamports on `accA` and
+  `Ix3` rebates 10 lamports on `accB`. In the case of multiple rebates only the
+  maximum of all the rebates is applied. Consider the transaction is executed
+  successfully. The maximum of all the rebates for `accA` is 75 lamports and
+  `accB` is 25 lamports. So a total of 100 lamports are rebated back to the
+  payer, `accA` gets 25 lamports and `accB` gets 75 lamports. The payer will end
+  up paying `other fees` + `100` lamports.
+
+* Fees paid full rebates but the execution failed case:
```

Comment:
Is this really the behaviour we want? If the program executed far enough that it decided the transaction creator earned a rebate, shouldn't it be honour even if the transaction later fails for some other completely unrelated reason? This seems more "fair" in a fully composable world. But I admittedly have probably not thought as hard as you have about this particular case

Reply:
Yes actually, if we do not want someone to add an instruction that fails at the end so they get the rebate and also maliciously use the program.

Reply:
Got it. Maliciously use in this case means "lock the account for write without paying the PRAW fee" right?

Reply:
Yeah